go/tools/gazelle/packages/fileinfo_proto.go

```diff
@@ -0,0 +1,133 @@
+/* Copyright 2017 The Bazel Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package packages
+
+import (
+	"bytes"
+	"io/ioutil"
+	"log"
+	"path"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/bazelbuild/rules_go/go/tools/gazelle/config"
+)
+
+var protoRe = buildProtoRegexp()
+
+const (
+	importSubexpIndex    = 1
+	packageSubexpIndex   = 2
+	goPackageSubexpIndex = 3
+	serviceSubexpIndex   = 4
+)
+
+func protoFileInfo(c *config.Config, dir, rel, name string) fileInfo {
+	info := fileNameInfo(dir, rel, name)
+	content, err := ioutil.ReadFile(info.path)
+	if err != nil {
+		log.Printf("%s: error reading proto file: %v", info.path, err)
+		return info
+	}
+
+	for _, match := range protoRe.FindAllSubmatch(content, -1) {
+		switch {
+		case match[importSubexpIndex] != nil:
+			imp := unquoteProtoString(match[importSubexpIndex])
+			info.imports = append(info.imports, imp)
+
+		case match[packageSubexpIndex] != nil:
+			pkg := string(match[packageSubexpIndex])
+			if info.packageName == "" {
+				info.packageName = strings.Replace(pkg, ".", "_", -1)
+			}
+
+		case match[goPackageSubexpIndex] != nil:
+			gopkg := unquoteProtoString(match[goPackageSubexpIndex])
+			if i := strings.LastIndexByte(gopkg, ';'); i != -1 {
+				info.importPath = gopkg[:i]
+				info.packageName = gopkg[i+1:]
+			} else {
+				info.importPath = gopkg
+				info.packageName = path.Base(gopkg)
+			}
+
+		case match[serviceSubexpIndex] != nil:
+			info.hasServices = true
+
+		default:
+			// Comment matched. Nothing to extract.
+		}
+	}
+	sort.Strings(info.imports)
+
+	if info.packageName == "" {
+		stem := strings.TrimSuffix(name, ".proto")
+		fs := strings.FieldsFunc(stem, func(r rune) bool {
+			return !('A' <= r && r <= 'Z' ||
```

Comment:
You probably ought to be using unicode.IsLetter and unicode.IsDigit in here.

Reply:
Done